### PR TITLE
First pass on the backend

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
The `FlightFileManager` backend is a fairly straight forward sinatra app. It requires a `Basic` `Authorization` header which is authenticated using the `sshd` pam module (the exact pam service is configurable).

There are three routes that have been implemented:

* `GET /ping` Checks if the service is running. Currently it is authenticated, but this will likely change
* `POST /cloudcmd` Creates a new `cloudcmd` session for the user
* `DELETE /cloudcmd` Destroy the existing `cloudcmd` session for the user

For full details on the routes see the [documentation](supervisor/docs/routes.md)

By default the create route runs `libexec/cloudcmd.sh` to launch the session. The first argument is the path to the pre-generated cloudcmd config file. The second argument is the `port_path` where the service has been launched. It is the responsibility of the script to select the port and write it to the `port_path`.

### CODE REVIEW

**Comment by benarmston** [comment]
Rebased on `main`.

File: supervisor/config/boot.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+require 'active_support/core_ext/hash/keys'
+
+lib = File.expand_path('../lib', __dir__)
+$LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
+
+require 'flight_file_manager'
+
+# Ensures the config has been loaded
+# XXX: Consider changing to a load method
+FlightFileManager.app.config
```
Comment (benarmston) [inline supervisor/config/boot.rb:55]: FYI, scheduler controller has `alias_method :load_configuration, :config` to handle this.  It does read a bit better.

**Review by benarmston — CHANGES_REQUESTED** [review]
Can you take a look at the comments/questions below.  I'll then merge to `main` and create an integration branch where I get this and the client working as I want.

File: supervisor/app.rb [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+        Process.getpgid(pid)
+        true
+      rescue Errno::ESRCH
+        return false
+      end
+    elsif File.exists?(pid_path)
+      status 500
+      halt({
+        errors: ['The internal state is inconsistent']
+      }.to_json)
```
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply supervisor/app.rb]: I initially had concerns what would happen in the `pid` was set without a `config_path`. However on review, the `config_path` is only used on create. Whilst it should continuing existing, it is not required.

I have removed the conditional and `config_path` check:
40e8a9dd05fc0124817317fea44f5cb464be2bb6

File: supervisor/app.rb [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+                      errors: ['A cloudcmd session already exists!'],
+                      port: File.read(port_path)
+                    )
+                    .to_json
+      status 409
+      halt payload
+    else
+      status 500
+      halt({
+        errors: ["The cloudcmd session for '#{current_user}' is missing its port! (PID: #{read_pid})"]
```
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply supervisor/app.rb:148]: 7095c1268fa9a597b030cf2b4b6c56af865131c8

File: supervisor/config/boot.rb (hunk 1 above)
Comment (WilliamMcCumstie) [reply supervisor/config/boot.rb]: 0846e773e5be0e3fa6b301044db75e394098d902

**Comment by benarmston** [comment]
Thanks for those changes @WilliamMcCumstie .

**Pull request merged**